### PR TITLE
Fix to #19609 - Query: allow user functions to be annotated with nullability propagation information

### LINKED ISSUE #19609: Query: allow user functions to be annotated with nullability propagation information
Related to #18555

We should allow users to specify nullability propagation info for their custom functions, so that we can generate better SQL when those functions are compared to null

### PR DESCRIPTION
**Pull request by maumar**
Added fluent API for function to specify it's nullability and function parameter to specify whether it propagates null. Also added property to DbFunction attribute to allow specify nullability there.

Fixes #19609

### CODE REVIEW

File: src/EFCore.Relational/Metadata/Builders/DbFunctionParameterBuilder.cs [hunk 1]
```diff
@@ -62,6 +62,17 @@ public virtual DbFunctionParameterBuilder HasStoreType([CanBeNull] string storeT
             return this;
         }
 
+        /// <summary>
+        ///     Indicates whether parameter propagates nullability, meaning if it's value is null the database function itself returns null.
+        /// </summary>
+        /// <param name="propagatesNullability"> Value which indicates whether parameter propagates nullability. </param>
+        public virtual DbFunctionParameterBuilder PropagatesNullability(bool propagatesNullability = true)
+        {
+            Builder.PropagatesNullability(propagatesNullability, ConfigurationSource.Explicit);
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Builders/DbFunctionParameterBuilder.cs:71]: missing return in docs

File: src/EFCore.Relational/Metadata/Builders/DbFunctionBuilderBase.cs [hunk 2]
```diff
@@ -86,6 +86,18 @@ public virtual DbFunctionBuilderBase IsBuiltIn(bool builtIn = true)
             return this;
         }
 
+        /// <summary>
+        ///     Marks whether the database function can return null value.
+        /// </summary>
+        /// <param name="nullable"> The value indicating whether the database function can return null. </param>
+        /// <returns> The same builder instance so that multiple configuration calls can be chained. </returns>
+        public virtual DbFunctionBuilderBase IsNullable(bool nullable = true)
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Builders/DbFunctionBuilderBase.cs:94]: This won't be here. It should only be defined for DbFunctionBuilder.
TableValuedFunctionBuilder does not have this attribute.

File: src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs [hunk 3]
```diff
@@ -82,17 +83,19 @@ public RelationalMethodCallTranslatorProvider([NotNull] RelationalMethodCallTran
                     return _sqlExpressionFactory.Function(
                         dbFunction.Name,
                         arguments,
-                        nullable: true,
-                        argumentsPropagateNullability: arguments.Select(a => false).ToList(),
+                        nullable: (dbFunction as DbFunction)?.IsNullable ?? true,
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs:86]: IDbFunction has IsNullable flag. Why soft cast?

Comment (maumar) [reply src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs:86]: forgo to remove after I added IsNullable on the interface.

**Comment by maumar** [comment]
updated

File: src/EFCore.Abstractions/DbFunctionAttribute.cs [hunk 4]
```diff
@@ -20,12 +20,14 @@ public class DbFunctionAttribute : Attribute
         private string _name;
         private string _schema;
         private bool _builtIn;
+        private bool _nullable;
```
Comment (AndriySvyryd) [inline src/EFCore.Abstractions/DbFunctionAttribute.cs:23]: To distinguish from explicit from default make this `bool?`, remove it from the constructor and add "GetIsNullable", see `IndexAttribute`

File: src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs [hunk 5]
```diff
@@ -82,17 +83,19 @@ public RelationalMethodCallTranslatorProvider([NotNull] RelationalMethodCallTran
...
 
                 return _sqlExpressionFactory.Function(
                     dbFunction.Schema,
                     dbFunction.Name,
                     arguments,
-                    nullable: true,
-                    argumentsPropagateNullability: arguments.Select(a => false).ToList(),
+                    nullable: dbFunction.IsNullable,
+                    argumentsPropagateNullability: dbFunction.Parameters.Select(p => p is DbFunctionParameter fp
+                        && fp.PropagatesNullability).ToList(),
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs:98]: Shouldn't we preserve which argument propagates nullability and which doesn't?

File: src/EFCore.Relational/Metadata/IStoreFunction.cs [hunk 6]
```diff
@@ -21,6 +21,11 @@ public interface IStoreFunction : ITableBase
         /// </summary>
         bool IsBuiltIn { get; }
 
+        /// <summary>
+        ///     Gets the value indicating whether the database function can return null value.
+        /// </summary>
+        bool IsNullable { get; }
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/IStoreFunction.cs:27]: I don't think this is needed

File: src/EFCore.Relational/Metadata/Internal/DbFunction.cs [hunk 7]
```diff
@@ -386,6 +390,40 @@ public virtual bool SetIsBuiltIn(bool builtIn, ConfigurationSource configuration
...
+
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public virtual bool SetIsNullable(bool nullable, ConfigurationSource configurationSource)
+        {
+            _nullable = nullable;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Internal/DbFunction.cs:413]: This should throw if the function isn't scalar

File: src/EFCore.Relational/Metadata/Internal/DbFunctionParameter.cs [hunk 8]
```diff
@@ -177,6 +179,50 @@ public virtual RelationalTypeMapping TypeMapping
...
+
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public virtual bool SetPropagatesNullability(bool propagatesNullability, ConfigurationSource configurationSource)
+        {
+            _propagatesNullability = propagatesNullability;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Internal/DbFunctionParameter.cs:202]: This should throw if the function isn't scalar

File: src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs (hunk 5 above)
Comment (maumar) [reply src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs:98]: we store the list with the assumptions the order of elements is the same as their order in the function definition, we then match it based on position in the list. Is that wrong assumption?

**Comment by maumar** [comment]
updated

File: src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs (hunk 5 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs:98]: nvm, I misread the `Select` as `Where`

Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs:98]: You don't need to call `ToList` though

File: src/EFCore.Relational/Metadata/Conventions/RelationalDbFunctionAttributeConvention.cs [hunk 9]
```diff
@@ -92,6 +92,11 @@ public virtual void ProcessModelFinalizing(IConventionModelBuilder modelBuilder,
                 {
                     dbFunctionBuilder.IsBuiltIn(dbFunctionAttribute.IsBuiltIn, fromDataAnnotation: true);
                 }
+
+                if (!dbFunctionAttribute.IsNullable)
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Metadata/Conventions/RelationalDbFunctionAttributeConvention.cs:96]: Use `GetIsNullable()` to only set it if not `null`

File: src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs (hunk 5 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/RelationalMethodCallTranslatorProvider.cs:98]: Also should not need cast to concrete type. It should be interface.

File: src/EFCore.Relational/Metadata/Conventions/RelationalDbFunctionAttributeConvention.cs [hunk 10]
```diff
@@ -92,6 +92,11 @@ public virtual void ProcessModelFinalizing(IConventionModelBuilder modelBuilder,
                 {
                     dbFunctionBuilder.IsBuiltIn(dbFunctionAttribute.IsBuiltIn, fromDataAnnotation: true);
                 }
+
+                if (!dbFunctionAttribute.GetIsNullable() != null)
+                {
+                    dbFunctionBuilder.IsNullable(dbFunctionAttribute.IsNullable, fromDataAnnotation: true);
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Conventions/RelationalDbFunctionAttributeConvention.cs:98]:
```suggestion
                if (dbFunctionAttribute.GetIsNullable() is bool value)
                {
                    dbFunctionBuilder.IsNullable(value, fromDataAnnotation: true);
```

File: src/EFCore.Relational/Metadata/Internal/DbFunction.cs [hunk 11]
```diff
@@ -113,6 +115,8 @@ public class DbFunction : ConventionAnnotatable, IMutableDbFunction, IConvention
                 : parameters
                     .Select(p => new DbFunctionParameter(this, p.Name, p.Type))
                     .ToList();
+
+            _nullable = true;
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/DbFunction.cs:119]: Only if it is scalar

File: src/EFCore.Relational/Metadata/Internal/DbFunctionParameter.cs [hunk 12]
```diff
@@ -177,6 +180,55 @@ public virtual RelationalTypeMapping TypeMapping
...
+            return propagatesNullability;
+        }
+
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        private void UpdatePropagatesNullabilityConfigurationSource(ConfigurationSource configurationSource)
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/DbFunctionParameter.cs:221]: You can inline this.

File: src/EFCore.Relational/Metadata/Internal/StoreFunction.cs [hunk 13]
```diff
@@ -50,6 +51,14 @@ public StoreFunction([NotNull] DbFunction dbFunction, [NotNull] RelationalModel
         /// <inheritdoc />
         public virtual bool IsBuiltIn { get; }
 
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public virtual bool IsNullable { get; }
```
Comment (smitpatel) [inline src/EFCore.Relational/Metadata/Internal/StoreFunction.cs:60]: Is this used anywhere?

File: test/EFCore.Relational.Specification.Tests/Query/UdfDbFunctionTestBase.cs [hunk 14]
```diff
@@ -820,6 +838,61 @@ public virtual void Nullable_navigation_property_access_preserves_schema_for_sql
...
+
+            Assert.Equal(4, result.Count);
+        }
+
+        [ConditionalFact]
+        public virtual void Compare_functions_returning_int_that_take_nullable_param_which_propagates_null()
+        {
+            using var context = CreateContext();
+
+            //var prm = default(string);
```
Comment (smitpatel) [inline test/EFCore.Relational.Specification.Tests/Query/UdfDbFunctionTestBase.cs:886]: remove

File: src/EFCore.Relational/Metadata/Internal/StoreFunction.cs (hunk 13 above)
Comment (maumar) [reply src/EFCore.Relational/Metadata/Internal/StoreFunction.cs:60]: added it everywhere just in case, will remove

**Comment by maumar** [comment]
updated @smitpatel

**Pull request merged**